### PR TITLE
entc/gen: make interceptor works for eager loading m2m edges

### PR DESCRIPTION
**Pull request by sisterdong**
This fixes: #3252

## What we expect

Take the following schema for example. 
image
First, we write a query to get all groups from the database.
```golang
groups, err := client.Group.Query().All(ctx)
if err != nil {
    return err
}
```
And then we add an interceptor to set a default behavior to GroupClient, which is always eager to load the groups' users and their pets.
```golang
client.Group.Intercept(ent.InterceptFunc(func(next ent.Querier) ent.Querier {
    return intercept.GroupFunc(func(ctx context.Context, query *ent.GroupQuery) (ent.Value, error) {
        query.WithUsers(func(q *UserQuery) { q.WithPets() })
        return next.Query(ctx, query)
    })
}))
```
We expect to get all the users from the groups, as well as each user's pets.

## What we have now
Interceptor doesn't work for Eager Loading M2M edges. 

1. Interceptors are only used in `All()`, `Count()` and `Scan()` query
https://github.com/ent/ent/blob/a85377846f75d58a21129aed9d28fedea7202387/entc/gen/template/builder/query.tmpl#L201-L209

2. All of the above three functions are not being used when generating code of eager loading M2M edges. But `All()` is used in other circumstances (like O2M) and works fine.
https://github.com/ent/ent/blob/a85377846f75d58a21129aed9d28fedea7202387/entc/gen/template/dialect/sql/query.tmpl#L98-L164

## How to solve this problem?
1.  Function`withInterceptors`  must be called in eager loading m2m edges, either through `All()`, `Count()`, `Scan()` or called directly.
2. Recursive eager loading should be supported, in the example above, this means pets should be correctly queried. Therefore, function `sqlAll` must be called, so we can't use `Scan()`.
https://github.com/ent/ent/blob/a85377846f75d58a21129aed9d28fedea7202387/entc/gen/template/dialect/sql/query.tmpl#L77-L84

I create this pr on the above ideas, I'd be glad if it makes some use.

### CODE REVIEW

**Comment by sisterdong** [comment]
@a8m Please take a look

**Review by a8m — COMMENTED** [review]
Thanks a lot for your contribution, @sisterdong.

File: entc/gen/template/dialect/sql/query.tmpl [hunk 1]
```diff
@@ -127,29 +127,32 @@ func ({{ $receiver }} *{{ $builder }}) sqlAll(ctx context.Context, hooks ...quer
...
+                        if nids[inValue] == nil {
+                            nids[inValue] = map[*{{ $.Name }}]struct{}{byID[outValue]: {}}
+                            return assign(columns[1:], values[1:])
+                        }
+                        nids[inValue][byID[outValue]] = struct{}{}
+                        return nil
+                    }
+                })
+            })
+			neighbors, err :=  withInterceptors[[]*{{ $e.Type.Name }}](ctx, query, qr, query.inters)
```
[the remark this answers is not in the document]
Comment (sisterdong) [reply entc/gen/template/dialect/sql/query.tmpl]: > Please, fix the indentation to make the review easier 🙏

Fixed.

**Review by a8m — APPROVED** [review]
Thanks for the contribution, @sisterdong.

**Comment by a8m** [comment]
Before merging this, can you run `go generate` on the `examples/` directory as well? Thanks 🙏

**Comment by sisterdong** [comment]
> Before merging this, can you run `go generate` on the `examples/` directory as well? Thanks 🙏

Of course. Done. It's my pleasure.

**Pull request merged**